src/main/scala/mesosphere/marathon/MarathonSchedulerActor.scala

```diff
@@ -446,17 +420,16 @@ class SchedulerActions(
         }.foreach { _ =>
           logger.info(s"Killing instances ${instances.map(_.instanceId)}")
 
-          def killInstances(instances: Seq[Instance]): Future[Done] = async {
+          async {
             val changeGoalsFuture = instances.map { i =>
-              if (i.hasReservation) instanceTracker.setGoal(i.instanceId, Goal.Stopped)
-              else instanceTracker.setGoal(i.instanceId, Goal.Decommissioned)
+              if (i.hasReservation) instanceTracker.setGoal(i.instanceId, Goal.Stopped, GoalAdjustmentReason.OverCapacity)
+              else instanceTracker.setGoal(i.instanceId, Goal.Decommissioned, GoalAdjustmentReason.OverCapacity)
             }
 
+            val instancesAreTerminal = killService.watchForKilledInstances(instances)
```

Comment:
Not sure I understand the suggestion – do you mean
```
decommission(instances: Seq[Instance], instancesKilled: Option[Promise[Done]])
OR
decommission(instances: Seq[Instance]): Option[Future[Done]])
```

?

I must say I don't really like either. Adjusting the goal to Stopped, Decommissioned or even Running doesn't tell us which status the call site is interested in. I believe there would be a default mapping like

```
Stopped -> all tasks are terminal
Decommissioned -> all tasks are terminal and associated reservations/volumes are destroyed
Running -> all tasks are Running
```

Now, for `Decommissioned` a call site might be OK with a terminal status (and assume the scheduler will eventually clean up). It might also be OK with `Unreachable`, since that could pertain for a while and there's probably no use in waiting for a definite terminal signal for the given use case.

For `Running`, a call site might be interested in `TASK_RUNNING` – but at the same time, e.g. during a deployment where we re-establish an instance that has been `Stopped` before, we might want to wait until the tasks are running, and health and readiness checks pass.

What I'm trying to say is: I would not try to imply one single success criterion for setting a goal. I can imagine something like
```
decommission(instance: Instance.Id, success: SuccessCriterion): Future[Done]

decommission("instance-123", TerminalTasks) // OK when all tasks are terminal
decommission("instance-123", InstanceExpunged) // OK when terminal and reservation cleaned up
decommission("instance-123", TerminalOrUnreachableTasks) // OK when all tasks are terminal or unreachable
```

But … not a fan.

Would you be OK if we rename or add some level of convenience? Otherwise this seems like something we can easily add afterwards.